File: dynamic_form_gen/src/main/java/com/ihsinformatics/dynamicformsgenerator/views/widgets/ScoreSpinner.java
package com.ihsinformatics.dynamicformsgenerator.views.widgets;

import android.content.Context;
import android.view.View;
import android.widget.AdapterView;
import android.widget.AdapterView.OnItemSelectedListener;
import android.widget.ArrayAdapter;
import android.widget.Spinner;
import android.widget.TextView;

import com.ihsinformatics.dynamicformsgenerator.R;
import com.ihsinformatics.dynamicformsgenerator.data.DataProvider;
import com.ihsinformatics.dynamicformsgenerator.data.Translator;
import com.ihsinformatics.dynamicformsgenerator.data.Translator.LANGUAGE;
import com.ihsinformatics.dynamicformsgenerator.data.core.options.Option;
import com.ihsinformatics.dynamicformsgenerator.data.core.questions.Question;
import com.ihsinformatics.dynamicformsgenerator.screens.BaseActivity;
import com.ihsinformatics.dynamicformsgenerator.utils.Global;
import com.ihsinformatics.dynamicformsgenerator.utils.GlobalPreferences;
import com.ihsinformatics.dynamicformsgenerator.utils.Validation;

import org.json.JSONException;
import org.json.JSONObject;

import java.util.ArrayList;
import java.util.HashMap;

public class ScoreSpinner extends InputWidget implements OnItemSelectedListener {

	TextView temp;
	private Spinner spAnswer;
	private ArrayAdapter<String> mAdapter;
	// ArrayList<String> mData;
	HashMap<String, String> mOptions;
	private int oldOptionScore;
	ArrayList<String> dataList;

	public ScoreSpinner(Context context, Question question, int layoutId) {
		super(context, question,layoutId);
		spAnswer = (Spinner) findViewById(R.id.spAnswer);
		// setBackgroundColor(getResources().getColor(android.R.color.transparent));
		spAnswer.setOnItemSelectedListener(this);
		spAnswer.setVisibility(View.GONE);
		findViewById(R.id.llQuestion).setVisibility(View.GONE);
		// if(options == null) options = DataProvider.getInstance(context).getOptions(question.getQuestionId());

		setOptionsOrHint(options.toArray(new Option[options.size()]));
	}
	
	@Override
	public boolean isValidInput(boolean isMendatory) {
		Validation validation = Validation.getInstance();
		return validation.validate(spAnswer, question.getValidationFunction(), isMendatory);
	}

	@Override
	public void setOptionsOrHint(Option... data) {
		if(data.length > 0) {
			dataList = new ArrayList<String>();
			mOptions = new HashMap<String, String>();
			
			for(int i=0; i<data.length; i++) {
				String translatedText = Translator.getInstance().Translate(data[i].getText(), GlobalPreferences.getinstance(context).findLanguagePrferenceValue());
				if(translatedText != null) {
					mOptions.put(translatedText, data[i].getText());
					dataList.add(translatedText);
				} else {
					mOptions.put(data[i].getText(), data[i].getText());
					dataList.add(data[i].getText());
					
				}
				
				
			}
			
			mAdapter = new ArrayAdapter<String>(context, android.R.layout.simple_list_item_1, dataList);
			spAnswer.setAdapter(mAdapter);
		}
	}
	
	public void addOption(String value) {
			mOptions.put(value, value);
			dataList.add(value);
			mAdapter.notifyDataSetChanged();
			spAnswer.setSelection(dataList.indexOf(value));
	}
	
	@Override
	public JSONObject getAnswer() throws JSONException {
		JSONObject param = new JSONObject();
		if (isValidInput(question.isMandatory())) {
			dismissMessage();
			if(spAnswer.getSelectedItem().toString().equals("<Select an option>")) {
                param.put(question.getParamName(), null);
			} else {
				LANGUAGE lang = GlobalPreferences.getinstance(context).findLanguagePrferenceValue();
				if (lang != LANGUAGE.ENGLISH) {
					String ans = mOptions.get(spAnswer.getSelectedItem().toString());
					param.put(question.getParamName(), ans);
				} else {
					param.put(question.getParamName(), spAnswer.getSelectedItem().toString());
				}
			}
			
		} else {
			activity.addValidationError(getQuestionId(), "Invalid input");
		}
		
		return param;
	}

	@Override
	public void onItemSelected(AdapterView<?> arg0, View arg1, int arg2, long arg3) {
        callOnItemSelectionChange();
		
	}
private void callOnItemSelectionChange() {
    String val = spAnswer.getSelectedItem().toString();
    int arg2 = spAnswer.getSelectedItemPosition();
    if(Global.SCOREABLE_QUESTIONS.contains(question.getQuestionId())) {
        Option o = findOptionByText(mOptions.get(val));
        if(o==null) {
            System.out.println("asd");
        }
        activity.addInRating(oldOptionScore-(oldOptionScore*2));
        activity.addInRating(o.getScore());
        oldOptionScore = o.getScore();
    }
    // TODO onItemClickListener can optimize the performance or try to make use of onNothingSelected
    try {
        int[] showables = options.get(arg2).getOpensQuestions();
        int[] hideables = options.get(arg2).getHidesQuestions();
        // if(getVisibility()==View.VISIBLE)
        ((BaseActivity)getContext()).onChildViewItemSelected(showables, hideables, question);

        // TODO disables this feature for CAD4TB project
			/*if("other".equals(val.toLowerCase(Locale.US))) {
				temp = (TextView) arg1;
				activity.startActivityForResult(new Intent(context, ManualInput.class), getQuestionId());
			}*/

    } catch(Exception e) {
        e.printStackTrace();
    }
}
	@Override
	public void onNothingSelected(AdapterView<?> arg0) {
		// TODO Auto-generated method stub
		
	}
	
	public void setOther(String value) {
		if(value == null) {
			// TODO make last selected item selected here in future
			spAnswer.setSelection(0);
			return;
		}
		try {
			if(!Validation.getInstance().areAllSpacesOrNull(value)){
				addOption(value);
				
			} else {
				// TODO make last selected item selected here in future
				spAnswer.setSelection(0);
			}
		} catch (Exception e) {
			e.printStackTrace();
		}
	}
	
	@Override
	public void setVisibility(int visibility) {
		if(spAnswer!=null) {
			// int[] showables = findAlldependantShowAbles();
			if(visibility == View.VISIBLE) {
				int[] showables = options.get(spAnswer.getSelectedItemPosition()).getOpensQuestions();
				int[] hideables = options.get(spAnswer.getSelectedItemPosition()).getHidesQuestions();
				((BaseActivity)getContext()).onChildViewItemSelected(showables, hideables, question);
			} else {
				int[] hideables = findAlldependantHideAbles();
				((BaseActivity)getContext()).onChildViewItemSelected(null, hideables, question);
			}
		}

		/*if(visibility == View.GONE) {
			if(Global.SCOREABLE_QUESTIONS.contains(question.getQuestionId())) {

				activity.addInRating(oldOptionScore-(oldOptionScore*2));
				activity.addInRating(0);
				// oldOptionScore = o.getScore();
			}
		} else {
			String val = spAnswer.getSelectedItem().toString();
			if(Global.SCOREABLE_QUESTIONS.contains(question.getQuestionId())) {
				Option o = findOptionByText(mOptions.get(val));
				if(o==null) {
					System.out.println("asd");
				}
				activity.addInRating(oldOptionScore-(oldOptionScore*2));
				activity.addInRating(o.getScore());
				oldOptionScore = o.getScore();
			}
		}*/
		
		super.setVisibility(visibility);
	}
	
	private int[] findAlldependantShowAbles() {
		int[] toReturn = new int[0];
		for(Option o: options) {
			int[] array = o.getOpensQuestions();
			if(array!=null) {
				int[] temp = toReturn;
				toReturn = new int[array.length+toReturn.length];
				
				for(int i=toReturn.length-array.length; i<toReturn.length; i++) {
					toReturn[i] = (array[i-temp.length]);
				}
				
				for(int i=0; i<temp.length; i++) {
					toReturn[i] = temp[i];
				}
			
			}
		}
		
		return toReturn;
	}
	
	private int[] findAlldependantHideAbles() {
		int[] toReturn = new int[0];
		for(Option o: options) {
			int[] array = o.getHidesQuestions();
			if(array!=null) {
				int[] temp = toReturn;
				toReturn = new int[array.length+toReturn.length];
				for(int i=toReturn.length-array.length; i<toReturn.length; i++) {
					toReturn[i] = (array[i-temp.length]);
				}
				for(int i=0; i<temp.length; i++) {
					toReturn[i] = temp[i];
				}
			}
		}
		
		return toReturn;
	}
	
	@Override
	public void setClickable(boolean clickable) {
		spAnswer.setClickable(clickable);
		super.setClickable(clickable);
	}
	
	public void setselectedIndex(int index) {
		spAnswer.setSelection(index);
        callOnItemSelectionChange();
	}
	
	@Override
	public void setEnabled(boolean enabled) {
		
		for(View lol : spAnswer.getTouchables() ) {
		    lol.setEnabled(enabled);
		}
		
		super.setEnabled(enabled);
	}

	@Override
	public void onFocusGained() {
		spAnswer.performClick();
		
	}
	
	@Override
	public void setAnswer(String answer, String uuid, LANGUAGE language) {
		answer = Translator.getInstance().Translate(answer, language);
		int i = dataList.indexOf(answer);
		if(i == -1) {
			addOption(answer);
			spAnswer.setSelection(dataList.indexOf(answer));
			return;
		}
		
		spAnswer.setSelection(dataList.indexOf(answer));
		setVisibility(View.VISIBLE);
		
	}

	private Option findOptionByText(String text) {
		for(Option o:options) {
			if(o.getText().equals(text)) {
				return o;
			}
		}
		
		return null;
	}

	@Override
	public void destroy() {
		// TODO Auto-generated method stub
		
	}
	@Override
	public String getValue() {
		return spAnswer.getSelectedItem().toString();
	}
	
}

File: app/src/main/java/ihsinformatics/com/hydra_mobile/ui/widgets/ScoreSpinner.kt
package ihsinformatics.com.hydra_mobile.ui.widgets

import android.content.Context
import android.view.View
import android.widget.AdapterView
import android.widget.AdapterView.OnItemSelectedListener
import android.widget.ArrayAdapter
import android.widget.Spinner
import android.widget.TextView
import ihsinformatics.com.hydra_mobile.R
import ihsinformatics.com.hydra_mobile.common.Constant
import ihsinformatics.com.hydra_mobile.data.core.options.Option
import ihsinformatics.com.hydra_mobile.data.core.question.Question
import ihsinformatics.com.hydra_mobile.utils.GlobalPreferences
import ihsinformatics.com.hydra_mobile.utils.Translator
import ihsinformatics.com.hydra_mobile.utils.Validation
import ihsinformatics.com.hydra_mobile.ui.base.BaseActivity
import org.json.JSONException
import org.json.JSONObject

import java.util.ArrayList
import java.util.HashMap

class ScoreSpinner(context: Context, question: Question, layoutId: Int) : InputWidget(context, question, layoutId),
    OnItemSelectedListener {

    private var temp: TextView? = null
    private val spAnswer: Spinner?
    private var mAdapter: ArrayAdapter<String>? = null
    private lateinit var mOptions: HashMap<String, String>
    private var oldOptionScore: Int = 0
    private lateinit var dataList: ArrayList<String>

    init {
        spAnswer = findViewById<View>(R.id.spAnswer) as Spinner
        // setBackgroundColor(getResources().getColor(android.R.color.transparent));
        spAnswer.onItemSelectedListener = this
        spAnswer.visibility = View.GONE
        findViewById<View>(R.id.llQuestion).visibility = View.GONE
        //Todo:handle the data provider class
        options = null//DataProvider.getInstance(context).getOptions(question.getQuestionId());

        setOptionsOrHint(*options.toTypedArray())
    }

    override fun isValidInput(isMendatory: Boolean): Boolean {
        val validation = Validation.getInstance()
        return validation.validate(spAnswer, question.validationFunction, isMendatory)
    }

    override fun setOptionsOrHint(vararg data: Option) {
        if (data.size > 0) {
            dataList = ArrayList()
            mOptions = HashMap()

            for (i in data.indices) {
                val translatedText = Translator.getInstance()
                    .Translate(data[i].text, GlobalPreferences.getinstance(context).findLanguagePrferenceValue())
                if (translatedText != null) {
                    mOptions[translatedText] = data[i].text
                    dataList.add(translatedText)
                } else {
                    mOptions[data[i].text] = data[i].text
                    dataList.add(data[i].text)

                }


            }

            mAdapter = ArrayAdapter(context, android.R.layout.simple_list_item_1, dataList)
            spAnswer!!.adapter = mAdapter
        }
    }

    fun addOption(value: String) {
        mOptions[value] = value
        dataList.add(value)
        mAdapter!!.notifyDataSetChanged()
        spAnswer!!.setSelection(dataList.indexOf(value))
    }

    @Throws(JSONException::class)
    override fun getAnswer(): JSONObject {
        val param = JSONObject()
        if (isValidInput(question.isMandatory)) {
            dismissMessage()
            if (spAnswer!!.selectedItem.toString() == "<Select an option>") {
                param.put(question.paramName, null)
            } else {
                val lang = GlobalPreferences.getinstance(context).findLanguagePrferenceValue()
                if (lang != Translator.LANGUAGE.ENGLISH) {
                    val ans = mOptions[spAnswer.selectedItem.toString()]
                    param.put(question.paramName, ans)
                } else {
                    param.put(question.paramName, spAnswer.selectedItem.toString())
                }
            }

        } else {
           // activity.addValidationError(questionId, "Invalid input")
        }

        return param
    }

    override fun onItemSelected(arg0: AdapterView<*>, arg1: View, arg2: Int, arg3: Long) {
        callOnItemSelectionChange()

    }

    private fun callOnItemSelectionChange() {
        val `val` = spAnswer!!.selectedItem.toString()
        val arg2 = spAnswer.selectedItemPosition
        if (Constant.SCOREABLE_QUESTIONS.contains(question.questionId)) {
            val o = findOptionByText(mOptions[`val`])
            if (o == null) {
                println("asd")
            }
           // activity.addInRating(oldOptionScore - oldOptionScore * 2)
           // activity.addInRating(o!!.score)
            //oldOptionScore = o.score
        }
        // TODO onItemClickListener can optimize the performance or try to make use of onNothingSelected
        try {
            val showables = options[arg2].opensQuestions
            val hideables = options[arg2].hidesQuestions
            // if(getVisibility()==View.VISIBLE)
           // (getContext() as BaseActivity).onChildViewItemSelected(showables, hideables)

            // TODO disables this feature for CAD4TB project
            /*if("other".equals(val.toLowerCase(Locale.US))) {
				temp = (TextView) arg1;
				activity.startActivityForResult(new Intent(context, ManualInput.class), getQuestionId());
			}*/

        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    override fun onNothingSelected(arg0: AdapterView<*>) {
        // TODO Auto-generated method stub

    }

    fun setOther(value: String?) {
        if (value == null) {
            // TODO make last selected item selected here in future
            spAnswer!!.setSelection(0)
            return
        }
        try {
            if (!Validation.getInstance().areAllSpacesOrNull(value)) {
                addOption(value)

            } else {
                // TODO make last selected item selected here in future
                spAnswer!!.setSelection(0)
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    override fun setVisibility(visibility: Int) {
        if (spAnswer != null) {
            // int[] showables = findAlldependantShowAbles();
            if (visibility == View.VISIBLE) {
                val showables = options[spAnswer.selectedItemPosition].opensQuestions
                val hideables = options[spAnswer.selectedItemPosition].hidesQuestions
              //  (getContext() as BaseActivity).onChildViewItemSelected(showables, hideables)
            } else {
                val hideables = findAlldependantHideAbles()
               // (getContext() as BaseActivity).onChildViewItemSelected(null, hideables)
            }
        }

        /*if(visibility == View.GONE) {
			if(Global.SCOREABLE_QUESTIONS.contains(question.getQuestionId())) {

				activity.addInRating(oldOptionScore-(oldOptionScore*2));
				activity.addInRating(0);
				// oldOptionScore = o.getScore();
			}
		} else {
			String val = spAnswer.getSelectedItem().toString();
			if(Global.SCOREABLE_QUESTIONS.contains(question.getQuestionId())) {
				Option o = findOptionByText(mOptions.get(val));
				if(o==null) {
					System.out.println("asd");
				}
				activity.addInRating(oldOptionScore-(oldOptionScore*2));
				activity.addInRating(o.getScore());
				oldOptionScore = o.getScore();
			}
		}*/

        super.setVisibility(visibility)
    }

    private fun findAlldependantShowAbles(): IntArray {
        var toReturn = IntArray(0)
        for (o in options) {
            val array = o.opensQuestions
            if (array != null) {
                val temp = toReturn
                toReturn = IntArray(array.size + toReturn.size)

                for (i in toReturn.size - array.size until toReturn.size) {
                    toReturn[i] = array[i - temp.size]
                }

                for (i in temp.indices) {
                    toReturn[i] = temp[i]
                }

            }
        }

        return toReturn
    }

    private fun findAlldependantHideAbles(): IntArray {
        var toReturn = IntArray(0)
        for (o in options) {
            val array = o.hidesQuestions
            if (array != null) {
                val temp = toReturn
                toReturn = IntArray(array.size + toReturn.size)
                for (i in toReturn.size - array.size until toReturn.size) {
                    toReturn[i] = array[i - temp.size]
                }
                for (i in temp.indices) {
                    toReturn[i] = temp[i]
                }
            }
        }

        return toReturn
    }

    override fun setClickable(clickable: Boolean) {
        spAnswer!!.isClickable = clickable
        super.setClickable(clickable)
    }

    fun setselectedIndex(index: Int) {
        spAnswer!!.setSelection(index)
        callOnItemSelectionChange()
    }

    override fun setEnabled(enabled: Boolean) {

        for (lol in spAnswer!!.touchables) {
            lol.isEnabled = enabled
        }

        super.setEnabled(enabled)
    }

    override fun onFocusGained() {
        spAnswer!!.performClick()

    }

    override fun setAnswer(answer: String, uuid: String, language: Translator.LANGUAGE) {
        var answer = answer
        answer = Translator.getInstance().Translate(answer, language)
        val i = dataList.indexOf(answer)
        if (i == -1) {
            addOption(answer)
            spAnswer!!.setSelection(dataList.indexOf(answer))
            return
        }

        spAnswer!!.setSelection(dataList.indexOf(answer))
        visibility = View.VISIBLE

    }

    private fun findOptionByText(text: String?): Option? {
        for (o in options) {
            if (o.text == text) {
                return o
            }
        }

        return null
    }

    override fun destroy() {
        // TODO Auto-generated method stub

    }


}
